### PR TITLE
Implement era1 file export

### PR DESCRIPTION
**Pull request by Matilda-Clerke**
## PR description
Implement a way to export ERA1 files

## Fixed Issue(s)
https://github.com/hyperledger/besu/issues/7935

### CODE REVIEW

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                            accumulatorHeaderRecords.add(
+                                new AccumulatorHeaderRecord(
+                                    block.getHash(), difficulty.toUInt256()));
+                          });
+                });
+      }
+      String network =
+          switch (blockchain.getGenesisBlockHeader().getHash().toString()) {
+            case MAINNET_GENESIS_HASH -> "mainnet";
+            case SEPOLIA_GENESIS_HASH -> "sepolia";
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: I thought this would work for any network, is there something specific that prevents this working in the general case? Don't recall the era1 import having any restrictions on networks

File: util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    this.hashFunction = new Digest();
+  }
+
+  /**
+   * Merkleize the supplied leaves as chunks
+   *
+   * @param leaves the leaves to merkleized
+   * @return the root hash from merkle tree produced by the supplied leaves
+   */
+  public Bytes32 merkleizeChunks(final List<Bytes32> leaves) {
```
Comment (jframe) [inline util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java:42]: Could we just use Tuweni merkleHash function instead https://github.com/Consensys/tuweni/blob/main/ssz/src/main/java/org/apache/tuweni/ssz/SSZ.java#L61?

File: util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+  /**
+   * Merkleize the supplied leaves as chunks, padding the leaves with zero chunks to padToSize if
+   * necessary
+   *
+   * @param leaves the leaves to merkleized
+   * @param padToSize the desired size to which leaves should be padded before merkleization
+   * @return the root hash from merkle tree produced by the supplied leaves
+   */
+  public Bytes32 merkleizeChunks(final List<Bytes32> leaves, final int padToSize) {
```
Comment (jframe) [inline util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java:54]: Whilst my preference would be reuse an existing like Tuweni that is already tested. If we it would be good to add some tests for these using the SSZ specs

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        blockchain
+            .getBlockByNumber(blockNumber)
+            .ifPresent(
+                (block) -> {
+                  blocksForFile.add(block);
+                  blockchain
+                      .getTxReceipts(block.getHash())
+                      .ifPresentOrElse(
+                          (receipts) -> transactionReceiptsForFile.put(block, receipts),
+                          () -> transactionReceiptsForFile.put(block, new ArrayList<>()));
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java:127]: Are the transaction receipts in the era1 file optional?

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  /**
+   * Exports ERA1 files starting from file startFile and ending at EndFile
+   *
+   * @param startFile The first ERA1 file to be exported
+   * @param endFile The last ERA1 file to be exported
+   * @param outputDirectory The directory in which to put the exported files
+   */
+  public void export(final long startFile, final long endFile, final File outputDirectory) {
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: Can you add a test for the export

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+  private long writeSection(
+      final FileOutputStream writer, final Era1Type era1Type, final byte[] content)
+      throws IOException {
+    byte[] typeCode = era1Type.getTypeCode();
+    writer.write(typeCode);
+    byte[] length = convertLengthToLittleEndianByteArray(content.length);
+    writer.write(length);
+    writer.write(content);
+
+    return (long) (typeCode.length + length.length + content.length);
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: nit: the cast isn't necessary here

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        snappyFactory.createFramedOutputStream(byteArrayOutputStream)) {
+      snappyOutputStream.write(content);
+    }
+
+    return writeSection(writer, era1Type, byteArrayOutputStream.toByteArray());
+  }
+
+  private byte[] convertLengthToLittleEndianByteArray(final int length) {
+    byte[] lengthBytes = Pack.intToLittleEndian(length);
+    return new byte[] {lengthBytes[0], lengthBytes[1], lengthBytes[2], lengthBytes[3], 0, 0};
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: why are 6 bytes being used to store the int?

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+      throw new IllegalArgumentException("End of export range must be after start of export range");
+    }
+    LOG.info("Exporting ERA1 files {} to {} inclusive", startFile, endFile);
+    for (long fileNumber = startFile; fileNumber <= endFile; fileNumber++) {
+      long startBlock = fileNumber * ERA1_FILE_BLOCKS;
+      long endBlock = startBlock + ERA1_FILE_BLOCKS - 1;
+
+      List<Block> blocksForFile = new ArrayList<>();
+      Map<Block, List<TransactionReceipt>> transactionReceiptsForFile = new HashMap<>();
+      Map<Block, Difficulty> difficultysForFile = new HashMap<>();
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]:
```suggestion
      Map<Block, Difficulty> difficultiesForFile = new HashMap<>();
```

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+                      .ifPresent(
+                          (difficulty) -> {
+                            difficultysForFile.put(block, difficulty);
+                            accumulatorHeaderRecords.add(
+                                new AccumulatorHeaderRecord(
+                                    block.getHash(), difficulty.toUInt256()));
+                          });
+                });
+      }
+      String network =
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: This doesn't need to be done on every iteration of the loop, the network shouldn't change

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java (hunk 1 above)
Comment (Matilda-Clerke) [reply app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: era1 files are specifically focused on pre-merge blocks. So post-merge sepolia/mainnet and other networks will be covered by a different format. Either the current era file, or the era-el file

File: util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java (hunk 2 above)
Comment (Matilda-Clerke) [reply util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java:42]: The tuweni ssz hash function is so out of date it uses 128 bit (16 byte) words instead of 256 bit (32 byte), so it's kind of useless for the current standard.

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java (hunk 4 above)
Comment (Matilda-Clerke) [reply app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java:127]: Not as far as I can tell, no.

File: util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java (hunk 3 above)
Comment (Matilda-Clerke) [reply util/src/main/java/org/hyperledger/besu/util/ssz/Merkleizer.java:54]: Good point, some tests are definitely in order

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java (hunk 6 above)
Comment (Matilda-Clerke) [reply app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: Frustratingly, gradle's compileJava takes issue with the raw ints being used in this context.
>/home/matilda/besu/app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java:215: >warning: [IntLongMath] Expression of type int may overflow before being assigned to a long
>      return typeCode.length + length.length + content.length;
>                                             ^
>      (see https://errorprone.info/bugpattern/IntLongMath)
>    Did you mean 'return (long) typeCode.length + length.length + content.length;'?

The correct fix is to cast each int to a long before adding them together, but it's quite unnecessary as the data is never expected to be that large.

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java (hunk 7 above)
Comment (Matilda-Clerke) [reply app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: This method is specifically for creating the byte array representing the length of the following section. The length is always 6 bytes and little endian.

File: app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java]: Now that there different formats we should make the `--format` visible

File: app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                  blockchain
+                      .getTotalDifficultyByHash(block.getHash())
+                      .ifPresent(
+                          (difficulty) -> {
+                            difficultysForFile.put(block, difficulty);
+                            accumulator.addBlock(block.getHash(), difficulty.toUInt256());
+                          });
+                });
+      }
+      String network =
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/chainexport/Era1BlockExporter.java]: You can get the network from the command line in the `BlocksSubCommand` and pass this in rather than inspecting the genesis file.

File: app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java [hunk 11]
```diff
@@ -401,6 +410,29 @@ private void exportRlpFormat(final BesuController controller) throws IOException
...
+      long startFile = getStartBlock().map(this::convertBlockNumberToFileNumber).orElse(0L);
+      long endFile =
+          getEndBlock()
+              .map(this::convertBlockNumberToFileNumber)
+              .orElse(
+                  convertBlockNumberToFileNumber(
+                      controller
+                          .getGenesisConfigOptions()
+                          .getCheckpointOptions()
+                          .getNumber()
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java]: This a brittle way to get the first POS block as the config for the checkpoint is not guaranteed to be the first Paris block although it currently is.

Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java]: I think the end block should also be limited to chain head as well so we don't try and export block we don't have.

File: app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java [hunk 12]
```diff
@@ -401,6 +410,29 @@ private void exportRlpFormat(final BesuController controller) throws IOException
...
+                          .orElseThrow()));
+
+      parentCommand
+          .era1BlockExporterFactory
+          .apply(controller.getProtocolContext().getBlockchain())
+          .export(startFile, endFile, blocksExportFile);
+    }
+
+    private long convertBlockNumberToFileNumber(final long blockNumber) {
```
Comment (jframe) [inline app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java]: Rather this conversion was done in the era exporter as the file number is a internal concern of the exporter. Everything else works in block numbers.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/rlp/BlockBodyRlpEncoder.java [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.core.rlp;
+
+import org.hyperledger.besu.ethereum.core.BlockBody;
+import org.hyperledger.besu.ethereum.rlp.RLP;
+
+import org.apache.tuweni.bytes.Bytes;
+
+public class BlockBodyRlpEncoder {
```
Comment (jframe) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/rlp/BlockBodyRlpEncoder.java]: I don't this or the other encoders in the rlp package belong in ethereum/core as they are not general purpose and quite specific to how the era1 export is encoding to RLP. Rather these were moved into the chain export package where they used.

File: util/src/test/java/org/hyperledger/besu/util/ssz/MerkleizerTest.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+import java.util.List;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.units.bigints.UInt256;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class MerkleizerTest {
```
Comment (jframe) [inline util/src/test/java/org/hyperledger/besu/util/ssz/MerkleizerTest.java:26]: Be good to test more edge cases. Such empty list, single chunk etc.

File: app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java (hunk 11 above)
Comment (Matilda-Clerke) [reply app/src/main/java/org/hyperledger/besu/cli/subcommands/blocks/BlocksSubCommand.java]: This is the method used elsewhere, which Simon and Gabriel settled on. Is there a better way? I originally just had it hardcoded in earlier history expiry work

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/rlp/BlockBodyRlpEncoder.java (hunk 13 above)
Comment (Matilda-Clerke) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/rlp/BlockBodyRlpEncoder.java]: I'm not sure I agree. These classes encode to RLP in fairly standard ways, which is also what is needed for the ERA1 export.

I'm happy to move them as requested, but that will limit their potential use. Let me know what you think.

Comment (jframe) [reply ethereum/core/src/main/java/org/hyperledger/besu/ethereum/core/rlp/BlockBodyRlpEncoder.java]: The reason I saw this as not being general purpose is that the `BlockBody.writeTo` is only used internally and by the `Era1BlockExporter`. And it also duplicates the encoder in the `BlockBody`.

Encoders are all in `org.hyperledger.besu.ethereum.core.encoding` and this and the other encoders is creating a parallel set of encoders.

Is the reason you added these so you can mock the encoding in the tests?

**Review by jframe — APPROVED** [review]
